src/solver.rs

```diff
@@ -246,20 +251,24 @@ pub trait DependencyProvider<P: Package, VS: VersionSet> {
     /// The kind of error returned from these methods.
     ///
     /// Returning this signals that resolution should fail with this error.
-    type Err: Error;
+    type Err: Error + 'static;
```

Comment:
What are the implications of the static here?

Reply:
That the error type must be able to outlive its dependency provider. It is less onerous than it might first appear. Nonetheless, I was surprised that I needed to add it. It's required by the `derive(Error)` generated code. I have not found a good way to look at the generated code to understand why it's needed. But I will give it another try.

Reply:
I used https://github.com/dtolnay/cargo-expand which worked really well. It would appear that `thiserror` decided that `Err` should be the [`source`](https://doc.rust-lang.org/std/error/trait.Error.html#method.source) of a `ErrorRetrievingDependencies`. Since `source` returns `dyn Error + 'static` we have to be `'static`. 

Looks like [`thiserror`](https://docs.rs/thiserror/latest/thiserror/) has special handling for fields named `source`. So we can:
- Remove the ` + 'static` and rename the field.
- Leave the ` + 'static` so that we can mark more things as `source`s.

I will add a commit doing the latter, but could go either way.